### ISSUE TITLE
Wrong inconsistency detected when overloading a function whose signature involves parametric types

### ISSUE THREAD

**Issue by jonathan-laurent** [microsoft/pyright#9054]
Pyright 1.1.381 sometimes raises spurious errors when overloading functions whose signature involves parametric types:

```py
# pyright: strict

from typing import overload

class Foo[T]:
    def method(self, x: T) -> int:
        assert False

@overload
def f(x: None) -> int:
    ...

@overload
def f[T](x: Foo[T]) -> int:
    ...

# Error with pyright v1.1.381
# (Accepted by mypy 1.11.1)
def f[T](x: Foo[T] | None) -> int:
    assert False
```

Here is the detailed error:

```
error: Overloaded implementation is not consistent with signature of overload 2
    Type "(x: Foo[T@f] | None) -> int" is not assignable to type "(x: Foo[T@f]) -> int"
      Parameter 1: type "Foo[T@f]" is incompatible with type "Foo[T@f] | None"
        Type "Foo[T@f]" is not assignable to type "Foo[T@f] | None"
          "notes.journal.2024-09-22-overload-bug.Foo" is not assignable to "notes.journal.2024-09-22-overload-bug.Foo"
            Type parameter "T@Foo" is contravariant, but "T@f" is not a supertype of "T@f"
          "Foo[T@f]" is not assignable to "None" (reportInconsistentOverload)
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.388.
